bundles/org.openhab.binding.gpio/src/main/java/org/openhab/binding/gpio/internal/handler/PigpioDigitalInputHandler.java

```diff
@@ -39,64 +40,176 @@
+        if (configuration.debouncingTime > 0) {
             scheduler.schedule(() -> afterDebounce(thisChange), configuration.debouncingTime, TimeUnit.MILLISECONDS);
-        });
-        jPigpio.gpioSetPullUpDown(gpio.getPin(), pullupdown);
+        } else {
+            afterDebounce(thisChange);
+        }
     }
 
+    /**
+     * Syncronize debouncing callbacks to
+     * ensure they are not out of order.
+     */
+    private Object debounceLock = new Object();
+
     private void afterDebounce(Date thisChange) {
-        try {
-            // Check if value changed over time
-            if (!thisChange.before(lastChanged)) {
-                updateStatus.accept(OnOffType.from(configuration.invert != gpio.getValue()));
+        synchronized (debounceLock) {
+            GPIO lgpio = this.gpio;
+            Consumer<State> lupdateStatus = this.updateStatus;
+
+            if (lgpio == null || lupdateStatus == null) {
+                // We raced and went offline in the meantime.
+                return;
+            }
+
+            try {
+                // Check if value changed over time
+                if (!thisChange.before(lastChanged)) {
+                    lupdateStatus.accept(OnOffType.from(configuration.invert != lgpio.getValue()));
+                }
+            } catch (PigpioException e) {
+                // -99999999 is communication related, we will let the Thing connect poll refresh it.
+                if (e.getErrorCode() != -99999999) {
+                    logger.error("Debounce exception :", e);
+                }
             }
+        }
+    }
+
+    /**
+     * Establishes or re-establishes a listener on the JPigpio
+     * instance for the configured gpio pin.
+     */
+    public void listen(@Nullable JPigpio jPigpio) throws PigpioException {
+        if (jPigpio == null) {
+            this.gpio = null;
+            return;
+        }
+
+        GPIO lgpio = new GPIO(jPigpio, this.gpioId, JPigpio.PI_INPUT);
+        this.gpio = lgpio;
+
+        try {
+            lgpio.setDirection(JPigpio.PI_INPUT);
+            jPigpio.gpioSetPullUpDown(lgpio.getPin(), this.pullupdown);
+            jPigpio.removeCallback(this.listener);
         } catch (PigpioException e) {
-            logger.warn("Unknown pigpio exception", e);
+            // If there is a communication error, the set alert below will throw.
+            if (e.getErrorCode() != -99999999) {
+                logger.error("Listen exception :", e);
+            }
         }
+
+        jPigpio.gpioSetAlertFunc(lgpio.getPin(), this.listener);
     }
 
     @Override
-    public void handleCommand(Command command) {
+    public void handleCommand(Command command) throws PigpioException {
+        GPIO lgpio = this.gpio;
+        Consumer<State> lupdateStatus = this.updateStatus;
+
+        if (lgpio == null || lupdateStatus == null) {
+            logger.warn("An attempt to submit a command was made when pigpiod was offline: {}", command.toString());
```

Comment:
see other comment, not sure if this should be changed to debug

Reply:
I think something above debug should be warranted here. The reason I had it at warn is due to the fact that this scenario represents a lost command. An action that the user requested could not be carried out. With debug, this would be normally suppressed and not known to the user. Note: The only valid command here is the REFRESH command. So it is most likely not a very common occurrence. Maybe info level would be better?

Reply:
I guess warn is fine then. Thanks